[Caffe2] Add support to TensorRT

yinghai opened the pull request:
Adding new PR as our C2 repo has merged into pytorch. Old PR at https://github.com/caffe2/caffe2/pull/2443. 

This PR adds an TensorRT operator and the graph cutting algorithm to cut a portion of "runnable" subgraph off input c2 net and off-load it to TensorRT to run. Some of the flow is still in Python but can be easily ported to C++.

To build:
```
cd build
cmake .. -DBUILD_SHARED_LIBS=OFF  -DBUILD_TEST=OFF -DBUILD_BENCHMARK=OFF -DBUILD_BINARY=OFF  -DUSE_CUDA=ON -DTENSORRT_ROOT=/opt/tensorrt -DUSE_TENSORRT=ON -DBUILD_CUSTOM_PROTOBUF=OFF
```

To test:
```
cd caffe2/python/trt
TEST_C2_TRT=1 pytest -v test_trt.py
```
TODO outside of this PR:
- Generic graph cutting/transforming algorithm with nomnigraph
- Workspace based initializer flow instead of relying on init_net
- On-the-fly RunNet conversion
- C++ version of SSA rewrite

Commit:
Add support to TensorRT

orionr commented on caffe2/contrib/tensorrt/tensorrt_op_trt.cc:
```diff
@@ -0,0 +1,240 @@
+/**
+ * Copyright (c) 2016-present, Facebook, Inc.
+ *
```
Can you remove this header? We don't put headers on the files anymore and rely on the base LICENSE file instead. This is PyTorch convention.

orionr commented on caffe2/contrib/tensorrt/tensorrt_op_trt.h:
```diff
@@ -0,0 +1,50 @@
+/**
+ * Copyright (c) 2016-present, Facebook, Inc.
```
Same here.

orionr reviewed (dismissed):
Please remove the license preambles. Thanks!

Commit:
Removed License header

yinghai commented:
@orionr Removed License comments
@Maratyszcza @dzhulgakov @Yangqing

Maratyszcza commented on caffe2/contrib/tensorrt/tensorrt_op_trt.cc:
```diff
@@ -0,0 +1,224 @@
+namespace {
+// Note that input of trt tensor is in CHW format, while our tensor is NCHW
+// \return -1 if there is dimension mismatch between C2 tensor and trt tensor.
+// Otherwise, return the multiplicaton of CHW dimensions
```
nit: `return the product of CHW dimensions`

Maratyszcza commented on caffe2/contrib/tensorrt/tensorrt_op_trt.cc:
```diff
@@ -0,0 +1,224 @@
+
+} // namespace
+
+// Upon construction, we build the inference enigne by deserializing from
```
typo: `enigne` -> `engine`

Maratyszcza commented on caffe2/contrib/tensorrt/tensorrt_op_trt.cc:
```diff
@@ -0,0 +1,224 @@
+TensorRTOp::TensorRTOp(const OperatorDef& operator_def, Workspace* ws)
+    : Operator<CUDAContext>(operator_def, ws),
+      logger_((nvinfer1::ILogger::Severity)(
+          OperatorBase::GetSingleArgument<int>("log_verbosity", 2))),
```
Please use `minloglevel` from here: https://github.com/pytorch/pytorch/blob/a2a28c0ef1d9a433972fe72fa5b0b9b850ccfcaf/caffe2/core/logging.cc#L106

yinghai replied:
I need to check `minloglevel` to see if their levels align well.

Maratyszcza commented on caffe2/contrib/tensorrt/tensorrt_op_trt.cc:
```diff
@@ -0,0 +1,224 @@
+    // TODO(support trt plugin factory)
+    trt_engine_ = InferObject(trt_runtime->deserializeCudaEngine(
+        engine_string.data(), engine_string.size(), nullptr));
+  }
```
Need a way to specify target device for inference (use `operator_def->device_option().cuda_gpu_id()`)

yinghai replied:
From the sample code, looks like it's set via plain `cudaSetDevice`. Maybe we can push/pop the old device id if it's not the same as `operator_def->device_option().cuda_gpu_id()` ?

dzhulgakov replied:
isn't it regular cudaSetDevice? I actually don't know how TRT exposes devices

yinghai replied:
Yes, it's the regular `CudaSetDevice`.

Maratyszcza commented on caffe2/contrib/tensorrt/tensorrt_op_trt.cc:
```diff
@@ -0,0 +1,224 @@
+    const auto it = outputs.find(oname);
+    if (it != outputs.end()) {
+      std::vector<TIndex> dims;
+      for (; idx < output_size_hints_encoded.size() && output_size_hints_encoded[idx] > 0; ++idx) {
```
Is there a use-case for specified, but non-positive output size hint? Otherwise, it is better to do `CAFFE_ENFORCE(output_size_hints_encoded[idx] > 0)`

yinghai replied:
I use `-1` as a delimiter here.

Maratyszcza commented on caffe2/contrib/tensorrt/tensorrt_op_trt.cc:
```diff
@@ -0,0 +1,224 @@
+  }
+
+  // match and bind the input/output
+  int num_bindings = trt_engine_->getNbBindings();
```
nit: `const int`

Maratyszcza commented on caffe2/contrib/tensorrt/tensorrt_op_trt.cc:
```diff
@@ -0,0 +1,224 @@
+    nv_dims_.push_back(trt_engine_->getBindingDimensions(b));
+    if (trt_engine_->bindingIsInput(b)) {
+      const auto it = inputs.find(name);
+      CAFFE_ENFORCE(it != inputs.end(), MakeString("Cannot find trt input: ", name));
```
You can do `CAFFE_ENFORCE(it != inputs.end(), "Cannot find trt input: ", name);` instead

Maratyszcza commented on caffe2/contrib/tensorrt/tensorrt_op_trt.cc:
```diff
@@ -0,0 +1,224 @@
+    } else {
+      const auto it = outputs.find(name);
+      CAFFE_ENFORCE(it != outputs.end());
+      binding_hints_.emplace_back(it->second, false);
```
BTW, both cases can be handles together, e.g.
```
const auto it = outputs.find(name);
CAFFE_ENFORCE(it != outputs.end());
binding_hints_.emplace_back(it->second, trt_engine_->bindingIsInput(b));
```

Maratyszcza commented on caffe2/contrib/tensorrt/tensorrt_op_trt.cc:
```diff
@@ -0,0 +1,224 @@
+    auto total_trt = std::accumulate(dims->begin(), dims->end(), (TIndex)(1), std::multiplies<TIndex>());
+    auto total_c2 = std::accumulate(dims_hint.begin(), dims_hint.end(), (TIndex)(1), std::multiplies<TIndex>());
+    if (total_c2 != total_trt) {
+      LOG(WARNING) << "The total size of TensorRT op output and hint don't match: " << total_trt << " vs " << total_c2;
```
Shouldn't it be a hard error if `total_c2 < total_trt`? In this case outputs will overflow Caffe2-allocated memory

Maratyszcza replied:
Nevermind, we allocate outputs before inference

yinghai replied:
Yes. I will error out.

yinghai replied:
Yes. Will revise.

Maratyszcza commented on caffe2/contrib/tensorrt/tensorrt_op_trt.cc:
```diff
@@ -0,0 +1,224 @@
+          identical_shape = false;
+          break;
+        }
+      }
```
Can replace this loop with `identical_shape = std::equal(dims->cbegin(), dims->cend(), dims_hint.cbegin());`

Maratyszcza commented on caffe2/contrib/tensorrt/tensorrt_op_trt.cc:
```diff
@@ -0,0 +1,224 @@
+  size_t N = 0;
+  bool first = true;
+  for (int i = 0; i < InputSize(); ++i) {
+    const auto& input_tensor = Input(i);
```
Please don't use `auto` here. It is non-obvious what is the return type.

yinghai replied:
I checked the code and using `auto` for well known function like `Input()` seems to be OK. https://github.com/pytorch/pytorch/blob/master/caffe2/core/operator.h#L67

Maratyszcza commented on caffe2/contrib/tensorrt/tensorrt_op_trt.cc:
```diff
@@ -0,0 +1,224 @@
+  bool first = true;
+  for (int i = 0; i < InputSize(); ++i) {
+    const auto& input_tensor = Input(i);
+    const auto& tensor_dims = input_tensor.dims();
```
You use `tensor_dims.front()`, probably need to check if there's more than one tensor dim (not sure if its guaranteed)

Maratyszcza commented on caffe2/contrib/tensorrt/tensorrt_op_trt.cc:
```diff
@@ -0,0 +1,224 @@
+  for (int i = 0; i < InputSize(); ++i) {
+    const auto& input_tensor = Input(i);
+    const auto& tensor_dims = input_tensor.dims();
+    if (first) {
```
You could use `if (i == 0)` and remove `first` variable.

Maratyszcza replied:
BTW, do you need to check that first dimension (batch size) does not exceed `batch_size_max` passed in OperatorDef? Or would TensorRT resize automatically?

Maratyszcza commented on caffe2/contrib/tensorrt/tensorrt_op_trt.cc:
```diff
@@ -0,0 +1,224 @@
+  // We need to do the binding at RunOnDevice time because we only know the
+  // exact shapes of the tensors now. In addtion, since TensorRT engine has
+  // max_batch_size, we need to call that multiple times if input batch size
+  // exceeeds this limit.
```
Could we avoid re-binding if shapes didn't change? A separate PR is fine

yinghai replied:
The tricky part is that we cannot tell if we need rebinding just by comparing the shapes. If the input batch size is larger than `max_batch_size` we need to do multiple bindings in one device run. So the bindings changes between op calls even if the shapes are the same.

yinghai replied:
I think binding is not costly. But if batch size exceeds `max_batch_size_` if makes sense to error out because calling the trt engine multiple times is not good.

Maratyszcza commented on caffe2/contrib/tensorrt/tensorrt_op_trt.cc:
```diff
@@ -0,0 +1,224 @@
+        offset + max_batch_size_ > N ? N - offset : max_batch_size_;
+    VLOG(2) << "Offset: " << offset << ", batch_size: " << batch_size << ", N: " << N;
+    int b = 0;
+    for (const auto& p : binding_hints_) {
```
Please don't use `auto` here: it is non-obvious what is the actual type

Maratyszcza commented on caffe2/contrib/tensorrt/tensorrt_op_trt.cc:
```diff
@@ -0,0 +1,224 @@
+    }
+
+    CAFFE_ENFORCE(bindings.size() == InputSize() + OutputSize());
+    if(!trt_executor_->execute(batch_size, &bindings[0])){
```
`&bindings[0]` -> `bindings.data()`

Maratyszcza commented on caffe2/contrib/tensorrt/tensorrt_op_trt.h:
```diff
@@ -0,0 +1,34 @@
+
+#include <NvInfer.h>
+#include <iostream>
+#include <memory>
```
`iostream` and `memory` doesn't seem needed for the interface file

yinghai replied:
I'll remove them and see. :)

Maratyszcza commented on caffe2/contrib/tensorrt/tensorrt_tranformer.cc:
```diff
@@ -0,0 +1,423 @@
+  return shape_hints;
+}
+
+std::vector<std::string> FigureInputs(
```
This function need better documentation. What is `start` and `end`? Why does it add op outputs to `boundary_inputs`?

yinghai replied:
`start` and `end` defines the continuous chunk of ops that can be readily converted into an TensorRT op. And this function tries to figure out what's the inputs of the to be converted TesnorRT op. What it does it collects the outputs from previous ops, which forms a cut boundary, and they can potential input of the TensorRT op (if referenced).

Maratyszcza commented on caffe2/contrib/tensorrt/tensorrt_tranformer.cc:
```diff
@@ -0,0 +1,423 @@
+          total_inputs_vec.emplace_back(input);
+          initialization_list->emplace(input);
+        } else if (boundary_inputs.count(input)) {
+          LOG(INFO) << "Adding boundary input: " << input;
```
VLOG seems more appropriate

dzhulgakov commented on caffe2/contrib/tensorrt/tensorrt_op_trt.cc:
```diff
@@ -0,0 +1,224 @@
+  for (size_t offset = 0; offset < N; offset += batch_size) {
+    bindings.clear();
+    batch_size =
+        offset + max_batch_size_ > N ? N - offset : max_batch_size_;
```
`min(N-offset, max_batch_size_)`

dzhulgakov replied:
nit: fix it?

dzhulgakov commented on caffe2/contrib/tensorrt/tensorrt_op_trt.cc:
```diff
@@ -0,0 +1,224 @@
+
+  std::unordered_map<std::string, int> inputs;
+  std::unordered_map<std::string, int> outputs;
+  for (int i = 0; i < operator_def.input_size(); ++i) {
```
I left a comment on the other repo - that it's better to not use blob names for bindings. Just use positions

Maratyszcza commented on caffe2/contrib/tensorrt/tensorrt_tranformer.cc:
```diff
@@ -0,0 +1,423 @@
+    const auto& op = pred_net.op(i);
+    for (const auto& output: op.output()) {
+      all_outputs.emplace(output);
+      all_outputs_vec.emplace_back(output);
```
You can have many duplicated in `all_outputs_vec`, which makes it inefficient. You should add only if `all_outputs` didn't have this element. Even more efficient would be to not have a vector in the first place, just iterate `all_outputs` and carefully use `std::unordered_set::erase`

yinghai replied:
Yes, you are right. I don't know why I did it this way. :( I'll changes.

dzhulgakov commented on caffe2/contrib/tensorrt/tensorrt_op_trt.cc:
```diff
@@ -0,0 +1,224 @@
+      return;
+    }
+
+    bool identical_shape = true;
```
just compare the vectors directly

dzhulgakov reviewed (changes requested):
some comments, will take more detailed view later

Maratyszcza commented on caffe2/contrib/tensorrt/tensorrt_tranformer.cc:
```diff
@@ -0,0 +1,423 @@
+  return all_outputs_vec;
+}
+
+std::vector<::ONNX_NAMESPACE::ValueInfoProto> ConvertToValueInfo(
```
I think it should be something different than `ONNX_NAMESPACE`, cc @bddppq

yinghai replied:
? lf it comes from onnx proto, it should have `ONNX_NAMESPACE`. On the other hand, if this is messed up, it won't compile. :)

Maratyszcza replied:
IIUC, `ONNX_NAMESPACE` should be replaced with something else during build.

Maratyszcza commented on caffe2/contrib/tensorrt/tensorrt_tranformer.cc:
```diff
@@ -0,0 +1,423 @@
+  return r;
+}
+
+void PruneWeights(const NetDef& pred_net, NetDef* init_net) {
```
A different name, e.g. `FindUsedWeights` would be better. Pruning of weights is already used in a different context (model compression)

yinghai replied:
OK

Maratyszcza commented on caffe2/contrib/tensorrt/tensorrt_tranformer.cc:
```diff
@@ -0,0 +1,423 @@
+    } else {
+      ++i;
+    }
+  }
```
This loop could be replaced with `std::remove_if`

yinghai replied:
Let me try. I remember trying some iterators with `RepeatedPtrField` but it doesn't seem to support it well.

yinghai replied:
I tried and it looks like old version of protobuf doesn't support `RepeatedPtrField::erase` with iterator.

Maratyszcza commented on caffe2/contrib/tensorrt/tensorrt_tranformer.cc:
```diff
@@ -0,0 +1,423 @@
+  }
+
+  auto engine_plan = InferObject(trt_engine->serialize());
+
```
`caffe2/utils/proto_utils.h` may be helpful here

yinghai replied:
@Maratyszcza  I actually met with a weird linkage error when I tried to use `MakeArgument` here. It complains  that `MakeArugument` symbol is missing during runtime. Any idea?

Maratyszcza commented on caffe2/contrib/tensorrt/tensorrt_tranformer.cc:
```diff
@@ -0,0 +1,423 @@
+    model->mutable_graph()->add_input()->CopyFrom(i);
+  }
+
+  // Convert weights to initializing tensors
```
This is going to be slow. Does TRT support any alternative?

yinghai replied:
Yes. Right now it is. But I think we can extend onnx-trt to support the initializeGraphIO once the backend spec is merged.

yinghai replied:
Sorry not initializeGraphIO but some alternatives. We do want to save copies from c2 to onnx to try.

Maratyszcza commented on caffe2/contrib/tensorrt/tensorrt_tranformer.cc:
```diff
@@ -0,0 +1,423 @@
+    int end,
+    const std::unordered_set<std::string>& weights,
+    const std::unordered_map<std::string, TensorShape>& shape_hints,
+    ::ONNX_NAMESPACE::ModelProto* model,
```
How is `model` supposed to be used? Why do you expect some nodes on input, but clear them before output?

yinghai replied:
Since onnx-trt takes in a model, we just use this model to wear all the ops for conversions. Since we want to reuse this model as a wrapper, we clear it after we finish the conversion.

Maratyszcza commented on caffe2/contrib/tensorrt/tensorrt_tranformer.cc:
```diff
@@ -0,0 +1,423 @@
+        if (!importer->supports(n)) {
+          LOG(INFO) << "TRT does not support ONNX node " << n.op_type();
+          support_trt = false;
+          break;
```
Why do you `break` here, but not in the check above?

yinghai replied:
Because they are not in the same loop. I'm breaking the inner loop here.

Maratyszcza commented on caffe2/contrib/tensorrt/tensorrt_tranformer.cc:
```diff
@@ -0,0 +1,423 @@
+        auto* node = onnx_model.mutable_graph()->add_node();
+        node->CopyFrom(n);
+      }
+    } else {
```
I don't understand the logic here. Why convert cluster to TRT can be successful if an op is not supported?

yinghai replied:
If we cannot convert some ops to trt op, due to 1. We cannot convert c2 op to onnx op
2. Trt backend doesn't support the onnx op
, It is still ok, we just leave that c2 op as is because we can still run it in c2.

Maratyszcza commented on caffe2/contrib/tensorrt/trt_utils.h:
```diff
@@ -0,0 +1,53 @@
+      Severity verbosity = Severity::kWARNING,
+      std::ostream& ostream = std::cout)
+      : _verbosity(verbosity), _ostream(&ostream) {}
+  void log(Severity severity, const char* msg) override {
```
Please forward it to Caffe2/GLOG logger instead, e.g. `LOG(INFO/WARNING/ERROR) << ...`

Maratyszcza commented on cmake/Dependencies.cmake:
```diff
@@ -587,3 +590,12 @@ list(APPEND Caffe2_DEPENDENCY_WHOLE_LINK_LIBS onnx_library)
+  include_directories("${PROJECT_SOURCE_DIR}/third_party/onnx-trt")
+  caffe2_interface_library(onnx2trt_importer_static onnx_trt_library)
+  list(APPEND Caffe2_DEPENDENCY_WHOLE_LINK_LIBS onnx_trt_library)
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -DUSE_TRT=1")
```
Don't change `CMAKE_CXX_FLAGS`, change `caffe2/core/macros.h.in` instead

Maratyszcza commented on cmake/public/cuda.cmake:
```diff
@@ -31,6 +31,23 @@ if(NOT CUDNN_FOUND)
 endif()
 
+# Optionally, find TensorRT
+if (${USE_TENSORRT} AND TENSORRT_ROOT)
```
Looks like you meant `AND DEFINED TENSORRT_ROOT`

Maratyszcza commented:
LGTM, some ideas for improvement:
1. Functions in `tensorrt_tranformer.cc` need better documentation. It is currently cryptic what many of them do and how.
2. I think referring to a subgraph by `start` and `end` nodes is suboptimal, as connected ops in a graph are not always representable by contiguous slice of operator list, and vice versa. `std::unordered_set` of node indexes would be a better to way to represent a subgraph.

dzhulgakov commented on caffe2/contrib/tensorrt/tensorrt_tranformer.cc:
```diff
@@ -0,0 +1,423 @@
+  auto importer = InferObject(onnx2trt::createImporter(trt_network.get()));
+  auto status = importer->import(onnx_model_str.data(), onnx_model_str.size(), false);
+  if (status.is_error()) {
+    CAFFE_THROW(MakeString(
```
CAFFE_THROW and other macros can take varargs to pass to MakeString directly

dzhulgakov commented on caffe2/contrib/tensorrt/tensorrt_tranformer.cc:
```diff
@@ -0,0 +1,423 @@
+
+  auto engine_plan = InferObject(trt_engine->serialize());
+
+  auto* serialized_engine_arg = op.add_arg();
```
yeah, please use proto_utils - MakeArgument and stuff like that

dzhulgakov commented on caffe2/contrib/tensorrt/tensorrt_tranformer.cc:
```diff
@@ -0,0 +1,423 @@
+    bool support_trt = true;
+    const OpSchema* schema = OpSchemaRegistry::Schema(op.type());
+    caffe2::onnx::ConvertedResult results;
+    if (!schema or schema->onnx_schema().empty()) {
```
`or` ? :)

yinghai replied:
Old habit from previous team. :)

dzhulgakov commented on caffe2/contrib/tensorrt/tensorrt_tranformer.cc:
```diff
@@ -0,0 +1,423 @@
+    } else {
+      // One c2 op can be converted into multiple onnx nodes. For simplicity, we
+      // enforce all or nothing here
+      results = onnx::OnnxExporter().Caffe2OpToOnnxNodes(op, shape_hints);
```
can this one throw (I didn't check)?

yinghai replied:
It doesn't.

dzhulgakov commented on caffe2/onnx/onnx_exporter.cc:
```diff
@@ -571,6 +571,67 @@ ConvertedResult OnnxExporter::CreateGemmNodes(
+void OnnxExporter::InitOpToTensorProto(
+    const caffe2::OperatorDef& op,
+    TensorProto* tensor) {
+  CAFFE_ENFORCE(op.input_size() == 0);
```
CAFFE_ENFORCE_EQ and stuff like that

dzhulgakov commented on caffe2/python/trt/transform.py:
```diff
@@ -0,0 +1,105 @@
+    """
+    Convert the whole ONNX model to a TensorRT C2 op
+    """
+    trt_str = C.onnx_to_trt_op(onnx_model.SerializeToString(),
```
so if compiled without TRT support or without GPU the method wouldn't just exist. So maybe produce a better messages in this case?

yinghai replied:
OK. I'll create stubs for this and error out with clear messages.

dzhulgakov commented on caffe2/python/trt/transform.py:
```diff
@@ -0,0 +1,105 @@
+    op.ParseFromString(trt_str)
+    return op
+
+def _infer_shapes(init_net, pred_net, inputs):
```
you can also try to invoke C2's shape inference (which is spotty though). See e.g. https://github.com/onnx/onnx-coreml/blob/4aa0b5a819418acc2495be629085f97f32e65208/onnx_coreml/_shapeinference.py#L31

yinghai replied:
Yeah, this is just a backup just in case C2's shape inference is spotty. I already used that.

dzhulgakov reviewed (changes requested):
more comments, mostly about clean ups. Please switch to positional parsing.

I didn't review the actual graph cutting logic in detail - as it might be replaced with nomnigraph in the future

Commit:
Bind input/output by position

yinghai commented on caffe2/contrib/tensorrt/tensorrt_op_trt.cc:
```diff
@@ -0,0 +1,224 @@
+    const auto it = outputs.find(oname);
+    if (it != outputs.end()) {
+      std::vector<TIndex> dims;
+      for (; idx < output_size_hints_encoded.size() && output_size_hints_encoded[idx] > 0; ++idx) {
```
I moved to using `output_size_hint_0`, `output_size_hint_1`.

yinghai commented on caffe2/contrib/tensorrt/tensorrt_op_trt.h:
```diff
@@ -0,0 +1,34 @@
+
+  TrtLogger logger_;
+  int max_batch_size_;
+  std::vector<std::pair<int, bool>> binding_hints_;
```
I'm removing this one as I moved to positional based binding.

Commit:
Comments

Commit:
More comments

yinghai commented:
Addressed most of the comments. Major revision includes:
- Use positional input/output binding
- More efficient algorithm for figuring out boundary inputs

The only remaining major issue is whether we should error out when batch size is larger than max batch size. I'll collect some numbers and see.

yinghai commented:
Regarding the `max_batch_size` I did some benchmark with Tesla M40 on my machine. Each test is doing inference 100 times after 20 times of warmup runs. The test next TensorRT engine is built with `max_batch_size=50`. Here are the results, regarding batch size. 
```
N=1, TRT op iter: 1
CUDA runtime: 1.07s
TRT runtime: 0.73s, improvement: 31.4%

N=50, TRT op iter: 1
CUDA runtime: 14.0s
TRT runtime: 8.30s, improvement: 41.6%

N=52, TRT op iter: 2
CUDA runtime: 14.4s
TRT runtime: 9.11s, improvement: 36.9%

N=101, TRT op iter: 3
CUDA runtime: 28.1s
TRT runtime: 17.3s, improvement: 38.2%
```

From the benchmark, looks like extra calls of inference engine has some cost (5% improvement drop from N=50 to N=52) but is still worth it.

Commit:
Add benchmark

Maratyszcza commented:
@yinghai Then it would be good to log a warning or info if batch is being split into multiple calls.

Commit:
Add warning for performance degradation on large batch

yinghai commented:
@onnxbot restest this please

yinghai commented:
@dzhulgakov  @Maratyszcza Any input/blockers?

dzhulgakov commented on caffe2/contrib/tensorrt/tensorrt_op_trt.cc:
```diff
@@ -0,0 +1,195 @@
+
+    CAFFE_ENFORCE_EQ(bindings.size(), InputSize() + OutputSize());
+    if(!trt_executor_->execute(batch_size, bindings.data())){
+      CAFFE_THROW("Error running the TensorRT executor");
```
does TRT have some reasonable error string?

yinghai replied:
It prints out info from Ilogger interface which we already provided.

dzhulgakov commented on caffe2/contrib/tensorrt/tensorrt_tranformer.h:
```diff
@@ -0,0 +1,53 @@
+#include <vector>
+
+namespace caffe2 {
+  class TensorRTTransformer {
```
nit - indentation. Run clang-format on the entire thing?

dzhulgakov commented on caffe2/contrib/tensorrt/trt_utils.h:
```diff
@@ -0,0 +1,48 @@
+};
+
+template <typename T>
+inline std::shared_ptr<T> InferObject(T* obj) {
```
should it be in another namespace? `InferObject` sounds cryptic

yinghai replied:
OK. I'll put it in a seperate namespace and rename the method.

dzhulgakov commented on caffe2/contrib/tensorrt/tensorrt_tranformer.cc:
```diff
@@ -0,0 +1,423 @@
+    } else {
+      end = op_idx;
+      ClusterToTrtOp(
+          *init_net,
```
simpler case for init_net might be to just create a throw-away Workspace and run the net. Then we don't need awkward `InitOpToTensorProto` and can even support more initializers directly. What do you think?

Even better option is to create this workspace at some higher level and make interface of conversion take workspace+predict_net instead of init_net+predict_net

dzhulgakov reviewed (commented):
Looking good, please fix some of the issues pointed out earlier (double-commented on them below)

yinghai commented on caffe2/contrib/tensorrt/tensorrt_op_trt.cc:
```diff
@@ -0,0 +1,195 @@
+  int64_t chw = 1;
+  for (int i = 0; i < nv_dims.nbDims; ++i) {
+    if (nv_dims.d[i] != c2_dims[i + 1]) {
+      return -1;
```
Sorry I missed this branch.

yinghai commented on caffe2/contrib/tensorrt/tensorrt_op_trt.cc:
```diff
@@ -0,0 +1,195 @@
+      return;
+    }
+
+    bool identical_shape = std::equal(dims->cbegin(), dims->cend(), dims_hint.cbegin());
```
I think you are right. We are enforcing the shape here anyway.

Commit:
Address comments

yinghai commented:
Addressed all the comments except for the workspace interface one, which I agree with the design but would like to put it in a separate diff to make things clear.

yinghai commented:
@pytorchbot restest this please

yinghai commented:
@pytorchbot retest this please

dzhulgakov reviewed (approved):
Cool, let's land it! (follow ups can be separate)

yinghai commented on caffe2/contrib/tensorrt/tensorrt_op_trt.cc:
```diff
@@ -0,0 +1,206 @@
+      // For output, we try to get its output size hint
+      const std::string key = MakeString("output_size_hint_", output_idx);
+      auto output_size_hint = OperatorBase::GetRepeatedArgument<int>(key);
+      if (!output_size_hint.empty()) {
```
`OperatorBase::GetRepeatedArgument` will return empty if argument doesn't exist. `output_size_hint` can be scalar. So here I think it's OK.

Commit:
comments

yinghai commented:
`pr/caffe2-conda2-ubuntu16.04-test` failure is unrelated.
